Add bad code attempt event to manual alarm control panel

giuseppeg88 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This Pull Request introduces a new event, `manual_alarm_bad_code_attempt`, for the `manual` alarm control panel integration. This event is fired whenever an attempt to change the state of a manual alarm (e.g., arm, disarm) fails due to an invalid code being provided.

The primary motivation for this change is to enhance the security and monitoring capabilities for users. By providing a dedicated event, users can now create automations to:
* Receive immediate notifications (e.g., via mobile app, email, or speaker) when an incorrect alarm code is entered.
* Log invalid code attempts for audit purposes.
* Trigger other actions (e.g., take a picture, turn on lights) in response to suspicious activity.

This new event provides the `entity_id` of the alarm, the `target state` that was attempted (e.g., `disarmed`, `armed_away`), and the `user_id` of the user who made the attempt (if available from the service call context).


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39426
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.


To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

giuseppeg88 committed:
manual: Fire bad code attempt event for invalid code attempts and add test

giuseppeg88 commented:
The request is still valid, waiting for an administrator to approve it

emontnemery committed:
Merge branch 'dev' into feature/manual-bad-code-event

emontnemery reviewed (commented):
Some comments.

emontnemery commented:
@giuseppeg88 Please address the review comments, after that this PR can be merged 👍

abmantis reviewed (changes requested):
I wonder if, instead of firing an event (which is a bad UX to automate), we could add a "failed_attempts" counter to the alarm entity attributes?

giuseppeg88 committed:
Refactor event listener for bad code attempts

Summary of the modifications:
- Registered the event listener on the Home Assistant event bus before making any service calls to ensure that all triggered events are captured.
- Checked that no "manual_alarm_bad_code_attempt" event is generated when using the correct alarm code during service calls.
- Verified that the "manual_alarm_bad_code_attempt" event is generated when an incorrect alarm code is used.
- Changed event capturing from a dictionary update to appending event data to a list, allowing multiple event captures rather than overwriting.

giuseppeg88 committed:
Change fire to async_fire for event handling

giuseppeg88 committed:
Change action to target_state in alarm control panel

giuseppeg88 committed:
Fix assertion in alarm control panel test

Update test to assert target_state instead of action.

giuseppeg88 committed:
Merge branch 'dev' into feature/manual-bad-code-event

giuseppeg88 reviewed (commented):
List of changes:
- Refactor event listener for bad code attempts
- Change fire to async_fire for event handling
- Change action to target_state in alarm control panel
- Fix assertion in alarm control panel test
- Merge branch 'dev' into feature/manual-bad-code-event

giuseppeg88 committed:
Merge branch 'dev' into feature/manual-bad-code-event

abmantis reviewed (changes requested):
Ruff is failing. Can you check it out?

giuseppeg88 committed:
Format code with Ruff

giuseppeg88 committed:
Merge branch 'dev' into feature/manual-bad-code-event
